std/range/primitives.d

```diff
@@ -1535,43 +1487,17 @@ static assert(isForwardRange!(typeof(r[1 .. 2])));
 static assert(hasLength!(typeof(r[1 .. 2])));
 ----
  */
-template hasSlicing(R)
-{
-    enum bool hasSlicing = isForwardRange!R && !isNarrowString!R && is(typeof(
-    (inout int = 0)
+enum bool hasSlicing(R) = isForwardRange!R
+    && !isNarrowString!R
+    && is(ReturnType!((R r) => r[1 .. 1].length) == size_t)
+    && (is(typeof(lvalueOf!R[1 .. 1]) == R) || isInfinite!R)
+    && (!is(typeof(lvalueOf!R[0 .. $])) || is(typeof(lvalueOf!R[0 .. $]) == R))
```

Comment:
I would replace tests like `!is(typeof(lvalueOf!R[0 .. $]))` with a more explicit `!hasMember!(R, "opDollar")`.

Same with `isRandomAccessRange`